### PR TITLE
Fix for persistent connection plugin on Python3

### LINKED ISSUE #24355: ios_command module doesn't work with Python3, raises TypeError
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
module - ios_command
plugin - connection/persistent.py

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 3.5.3 |Continuum Analytics, Inc.| (default, Mar  6 2017, 12:15:08) [GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
```

##### CONFIGURATION

default configuration

##### OS / ENVIRONMENT

OSX 10.12.4, conda 4.3.17, Python 3.5.3
Emulating Cisco routers with GNS3 1.5.3, Cisco IOS 15.2(4)S5

##### SUMMARY
ios_command module doesn't work with Python 3.5. Same playbook works with Python 2.7.
Error is "TypeError: string argument expected, got 'bytes'."
Can be Python 3 compatibility issue.
SSH debug on the routers does not show any connection attempts.

##### STEPS TO REPRODUCE

Run the playbook with the following command:
`ansible-playbook -i ~/prog/Ansible/Cisco-IOS-test/inventory ~/prog/Ansible/Cisco-IOS-test/sh-clock.yml -vvv`
(file paths are locally significant, shown as example).
inventory file:
```
# Example inventory file:
[ios]
10.0.15.1
10.0.15.2
10.0.15.3
10.0.15.4
```

```yaml
hosts: ios
  gather_facts: no
  connection: local

  tasks:

  - name: Credentials
    include_vars: ~/prog/Ansible/Cisco-IOS-test/creds.yml

  - name: DEFINE PROVIDER
    set_fact:
      provider:
        port: 22
        username: "{{ creds['username'] }}"
        password: "{{ creds['password'] }}"
        auth_pass: "{{ creds['auth_pass'] }}"

  - name: Show clock
    ios_command:
      provider: "{{ provider }}"
      commands: 
       - "show clock"
    register: clock

  - debug: msg={{ clock.stdout }}
```

##### EXPECTED RESULTS

Python 2.7 output:
```
(ansibleenv) MacBook-Pro:~ sbondar$ ansible-playbook -i ~/prog/Ansible/Cisco-IOS-test/inventory ~/prog/Ansible/Cisco-IOS-test/sh-clock.yml

PLAY [ios] ********************************************************************************************************************

TASK [Credentials] ************************************************************************************************************
ok: [10.0.15.4]
ok: [10.0.15.3]
ok: [10.0.15.2]
ok: [10.0.15.1]

TASK [DEFINE PROVIDER] ********************************************************************************************************
ok: [10.0.15.1]
ok: [10.0.15.4]
ok: [10.0.15.3]
ok: [10.0.15.2]

TASK [Show clock] *************************************************************************************************************
ok: [10.0.15.2]
ok: [10.0.15.1]
ok: [10.0.15.3]
ok: [10.0.15.4]

TASK [debug] ******************************************************************************************************************
ok: [10.0.15.2] => {
    "changed": false, 
    "msg": [
        "*04:33:24.756 UTC Tue Apr 25 2017"
    ]
}
ok: [10.0.15.1] => {
    "changed": false, 
    "msg": [
        "*05:14:06.220 UTC Tue Apr 25 2017"
    ]
}
ok: [10.0.15.3] => {
    "changed": false, 
    "msg": [
        "*04:32:48.352 UTC Tue Apr 25 2017"
    ]
}
ok: [10.0.15.4] => {
    "changed": false, 
    "msg": [
        "*04:32:48.648 UTC Tue Apr 25 2017"
    ]
}

PLAY RECAP ********************************************************************************************************************
10.0.15.1                  : ok=4    changed=0    unreachable=0    failed=0   
10.0.15.2                  : ok=4    changed=0    unreachable=0    failed=0   
10.0.15.3                  : ok=4    changed=0    unreachable=0    failed=0   
10.0.15.4                  : ok=4    changed=0    unreachable=0    failed=0   
```

##### ACTUAL RESULTS

Python 3.5 output with error:

```
(ansibleenv4) MacBook-Pro:~ sbondar$ ansible-playbook -i ~/prog/Ansible/Cisco-IOS-test/inventory ~/prog/Ansible/Cisco-IOS-test/sh-clock.yml -vvvv
No config file found; using defaults
Loading callback plugin default of type stdout, v2.0 from /Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/callback/__init__.py

PLAYBOOK: sh-clock.yml ********************************************************************************************************
1 plays in /Users/sbondar/prog/Ansible/Cisco-IOS-test/sh-clock.yml

PLAY [ios] ********************************************************************************************************************
META: ran handlers

TASK [Credentials] ************************************************************************************************************
task path: /Users/sbondar/prog/Ansible/Cisco-IOS-test/sh-clock.yml:8
ok: [10.0.15.1] => {
    "ansible_facts": {
        "creds": {
            "auth_pass": "cisco",
            "password": "cisco",
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.2] => {
    "ansible_facts": {
        "creds": {
            "auth_pass": "cisco",
            "password": "cisco",
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.3] => {
    "ansible_facts": {
        "creds": {
            "auth_pass": "cisco",
            "password": "cisco",
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.4] => {
    "ansible_facts": {
        "creds": {
            "auth_pass": "cisco",
            "password": "cisco",
            "username": "cisco"
        }
    },
    "changed": false
}

TASK [DEFINE PROVIDER] ********************************************************************************************************
task path: /Users/sbondar/prog/Ansible/Cisco-IOS-test/sh-clock.yml:11
ok: [10.0.15.2] => {
    "ansible_facts": {
        "provider": {
            "auth_pass": "cisco",
            "password": "cisco",
            "port": 22,
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.1] => {
    "ansible_facts": {
        "provider": {
            "auth_pass": "cisco",
            "password": "cisco",
            "port": 22,
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.3] => {
    "ansible_facts": {
        "provider": {
            "auth_pass": "cisco",
            "password": "cisco",
            "port": 22,
            "username": "cisco"
        }
    },
    "changed": false
}
ok: [10.0.15.4] => {
    "ansible_facts": {
        "provider": {
            "auth_pass": "cisco",
            "password": "cisco",
            "port": 22,
            "username": "cisco"
        }
    },
    "changed": false
}

TASK [Show clock] *************************************************************************************************************
task path: /Users/sbondar/prog/Ansible/Cisco-IOS-test/sh-clock.yml:19
The full traceback is:
Traceback (most recent call last):
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 521, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/action/ios.py", line 69, in run
    rc, out, err = connection.exec_command('open_shell()')
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 70, in exec_command
    return self._do_it('EXEC: ' + cmd)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 56, in _do_it
    cPickle.dump(self._play_context.serialize(), src)
TypeError: string argument expected, got 'bytes'

fatal: [10.0.15.2]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
The full traceback is:
Traceback (most recent call last):
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 521, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/action/ios.py", line 69, in run
    rc, out, err = connection.exec_command('open_shell()')
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 70, in exec_command
    return self._do_it('EXEC: ' + cmd)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 56, in _do_it
    cPickle.dump(self._play_context.serialize(), src)
TypeError: string argument expected, got 'bytes'

fatal: [10.0.15.3]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
The full traceback is:
Traceback (most recent call last):
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 521, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/action/ios.py", line 69, in run
    rc, out, err = connection.exec_command('open_shell()')
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 70, in exec_command
    return self._do_it('EXEC: ' + cmd)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 56, in _do_it
    cPickle.dump(self._play_context.serialize(), src)
TypeError: string argument expected, got 'bytes'

fatal: [10.0.15.1]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
The full traceback is:
Traceback (most recent call last):
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/executor/task_executor.py", line 521, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/action/ios.py", line 69, in run
    rc, out, err = connection.exec_command('open_shell()')
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 70, in exec_command
    return self._do_it('EXEC: ' + cmd)
  File "/Users/sbondar/anaconda/envs/ansibleenv4/lib/python3.5/site-packages/ansible/plugins/connection/persistent.py", line 56, in _do_it
    cPickle.dump(self._play_context.serialize(), src)
TypeError: string argument expected, got 'bytes'

fatal: [10.0.15.4]: FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
	to retry, use: --limit @/Users/sbondar/prog/Ansible/Cisco-IOS-test/sh-clock.retry

PLAY RECAP ********************************************************************************************************************
10.0.15.1                  : ok=2    changed=0    unreachable=0    failed=1   
10.0.15.2                  : ok=2    changed=0    unreachable=0    failed=1   
10.0.15.3                  : ok=2    changed=0    unreachable=0    failed=1   
10.0.15.4                  : ok=2    changed=0    unreachable=0    failed=1   

```

### PR DESCRIPTION
**Pull request by abadger**
On Python3, pickle needs to work with byte strings, not text strings.
Modify the code to do that.

Fixes #24355

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
plugins/connection/persistent.py

##### ANSIBLE VERSION

```
devel 2.3
```

### CODE REVIEW

**Comment by ganeshrn** [comment]
After applying the fix, I am getting this error on py3
```
<ios01> using connection plugin network_cli
<ios01> socket_path: /Users/gnalawad/.ansible/pc/e2219bf102
open_shell() returned 1 b'' b'Traceback (most recent call last):\n  File "/Users/gnalawad/ansible/ganeshrn/ansible/bin/ansible-connection", line 268, in main\n    raise Exception("EOL found before init data was complete")\nException: EOL found before init data was complete\nFAIL: EOL found before init data was complete\n'
fatal: [ios01]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell",
    "rc": 1
}
```

~~I modify below line to make it work, tested on py2.7 and py3.5 and it works fine~~
I did mistake while testing, activated py3.5 environment and ran playbook while background process from py2.7 was still active. 
So `protocol=0` work on py2.7 and not on Py3.5
```
- src = cPickle.dumps(self._play_context.serialize()
+ src = cPickle.dumps(self._play_context.serialize(), protocol=0)
```
Also had to comment ```src.close()``` line as it was throwing execption.
```
AttributeError: 'bytes' object has no attribute 'close'
```

File: lib/ansible/plugins/connection/persistent.py [hunk 1]
```diff
@@ -52,9 +52,8 @@ def _do_it(self, action):
         stdin = os.fdopen(master, 'wb', 0)
         os.close(slave)
 
-        src = StringIO()
-        cPickle.dump(self._play_context.serialize(), src)
-        stdin.write(src.getvalue())
+        src = cPickle.dumps(self._play_context.serialize())
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/persistent.py]: By default pickle protocol for `PY3` is `3` which is not backward compatible. Need to pass `protocol` value to make it backward compatible.

File: lib/ansible/plugins/connection/persistent.py [hunk 2]
```diff
@@ -52,9 +52,8 @@ def _do_it(self, action):
...
+        stdin.write(src)
         src.close()
```
Comment (ganeshrn) [inline lib/ansible/plugins/connection/persistent.py]: As `cPickle.dumps` returns a byte object this line is not required.

**Comment by abadger** [comment]
@ganeshrn Interesting.  I'm guessing that what's happening is that python3 is being used locally (which uses protocol version 3 by default) but remotely python2 is being used (which only supports protocol versions 0, 1, and 2).  protocol version=2 would be more efficient and supported by all Python versions we support.  Does that work? (I'll update the PR to use protocol=2 in a moment).

@jimi-c, where is this pickle loaded in the end?  Looking at the pickle docs in python3, I think we might need to change the encoding parameter to pickle.load() to be encoding='bytes' when we're on python3 (on python2, pickle.load() doesn't have an encoding parameter).

**Comment by abadger** [comment]
I think i've found where pickle.loads() is used: in bin/ansible-connection.  Thinking about the effect of changing encoding... I think that encoding='bytes' is the right thing to do but am a little worried about what may happen if we're not properly decoding all values in  the serialized play_context to text.  If we end up running the controller with python2 and ansible-conenction with python3 and we don't properly decode all values to text then ansible-connection will suddenly start getting some bytes values instead of text values with the encoding=bytes change.  That could lead to tracebacks.  Because of that, I think that I should make a separate PR for that and perhaps only push that change to devel.

**Comment by abadger** [comment]
@ganeshrn also reported to me that protocol=2 and protocol=1 still errors:

for PY2.7

```<ios01> using connection plugin network_cli
<ios01> socket_path: /Users/gnalawad/.ansible/pc/e2219bf102
open_shell() returned 1  Traceback (most recent call last):
  File "/Users/gnalawad/ansible/ganeshrn/ansible/bin/ansible-connection", line 271, in main
    pc_data = cPickle.load(src)
UnpicklingError: could not find MARK
FAIL: could not find MARK
fatal: [ios01]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell ",
    "rc": 1
}
```
and on PY3.5

```<ios01> using connection plugin network_cli
<ios01> socket_path: /Users/gnalawad/.ansible/pc/e2219bf102
open_shell() returned 1 b'' b'Traceback (most recent call last):\n  File "/Users/gnalawad/ansible/ganeshrn/ansible/bin/ansible-connection", line 267, in main\n    raise Exception("EOL found before init data was complete")\nException: EOL found before init data was complete\nFAIL: EOL found before init data was complete\n'
fatal: [ios01]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell ",
    "rc": 1
}
```

I'll need a reproducer to diagnose why only protocol=1 works.

**Comment by ganeshrn** [comment]
@abadger Updated my earlier comment. 
`protocol=0` work on py2.7 and not on py3.5

**Comment by abadger** [comment]
Okay, pushed some more changes which makes the test case "work" for me.  (I don't have a device to connect to so "work" means that it gets to the point in the code where it fails to connect to the remote device.  The changes to ansible-connection weren't needed for this bug report (because we're still running ansible-connection under python2 in this test case) but will be needed in setups that change its shebang to python3 (Probably what occurs if you python setup.py install ansible).  The changes to persistant.py are: (1) To switch to protocol=0.  This turns out to be because we're using a pty here, not simply piping to the process's stdin.  The pty interprets control chars so we have to be careful not to write arbitrary bytes to the process.  (2) To move closing of the process's stdin to after we run it.  For some reason you can close it before running on Python2 but on Python3, that means that nothing is sent to the process.

@ganeshrn or anyone else give the newest version a try and see if it works now.

**Comment by tjuberg** [comment]
The assumption that ansible-connection will still run under python2 is a fallacy since Arch and others have /usr/bin/python as python3, not python2 as these distributions moved to python3 prior to the Python project providing recommendations that /usr/bin/python should point to python2.

As such this is the result for me on Arch:
```
2017-05-11 21:14:11,732 p=22777 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9 
2017-05-11 21:14:12,380 p=22777 u=tjuberg |  open_session() returned 1 b'' b'Traceback (most recent call last):\n  File "/home/tjuberg/.local/badger-ansible/bin/ansible-connection", line 274, in main\n    pc_data = cPickle.load(init_data)\nTypeError: file must have \'read\' and \'readline\' attributes\nFAIL: file must have \'read\' and \'readline\' attributes\n' 
2017-05-11 21:14:12,390 p=22777 u=tjuberg |  fatal: [trd-gw]: FAILED! => { 
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 1 
} 

```

**Comment by abadger** [comment]
@TJuberg <nod> Arch has long been known to suffer from that problem.  The workaround is to use /usr/bin/python2 explicitly when running setup.py so that the python2 interpreter is substituted into the shebang.

that said, I've been working on fixing the python3 incompatiblities, I just don't directly have access to test all functionalities.  So as long as you're willing to keep testing the code with python3, I'm willing to keep fixing the bugs you find :-)

The one you just had was a typo introduced by this PR.  I've pushed a new commit, give it a try and tell me what else you find.

(You may also be interested in https://github.com/ansible/ansible/pull/24454/files which I'll rebase on top of this PR shortly.)

**Comment by tjuberg** [comment]
@badger I'll be happy to assist in any way I can while I am still working on my bachelors project and have access to the equipment for testing.

I'll see about setting up some docker containers for testing both on the special case of Arch with Python3 as well as the more common scenario with ansible_python_interpreter = /usr/bin/python3 on Ubuntu.

With the latest commits:

```
2017-05-12 08:46:03,521 p=6697 u=tjuberg |  Using /home/tjuberg/work/gitlab/bachelor_ansible/ansible.cfg as config file 
2017-05-12 08:46:03,722 p=6697 u=tjuberg |  Loading callback plugin default of type stdout, v2.0 from /home/tjuberg/.local/badger-ansible/lib/ansible/plugins/callback/__init__.py 
2017-05-12 08:46:03,777 p=6697 u=tjuberg |  PLAYBOOK: junos_facts.yml ****************************************************************************************************** 
2017-05-12 08:46:03,777 p=6697 u=tjuberg |  1 plays in junos_facts.yml 
2017-05-12 08:46:03,811 p=6697 u=tjuberg |  PLAY [Get JunOS facts] ********************************************************************************************************* 
2017-05-12 08:46:03,825 p=6697 u=tjuberg |  META: ran handlers 
2017-05-12 08:46:03,827 p=6697 u=tjuberg |  TASK [Gather Junos facts] ****************************************************************************************************** 
2017-05-12 08:46:03,827 p=6697 u=tjuberg |  task path: /home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.yml:9 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  creating new control socket for host xxx.xxx.xxx.xxx:22 as user ansible 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  control socket path is /home/tjuberg/.ansible/pc/d930766c35 
2017-05-12 08:46:04,428 p=6706 u=tjuberg |  current working directory is /home/tjuberg/work/gitlab/bachelor_ansible 
2017-05-12 08:46:04,429 p=6706 u=tjuberg |  using connection plugin netconf 
2017-05-12 08:46:04,530 p=6706 u=tjuberg |  network_os is set to junos 
2017-05-12 08:46:04,530 p=6706 u=tjuberg |  ssh connection done, stating ncclient 
2017-05-12 08:46:04,544 ncclient.transport.ssh Connected (version 2.0, client OpenSSH_6.6.1) 
2017-05-12 08:46:04,866 ncclient.transport.ssh Authentication (password) successful! 
2017-05-12 08:46:05,019 ncclient.transport.session initialized: session-id=19666 | server_capabilities=<dict_keyiterator object at 0x7f348c13ef48> 
2017-05-12 08:46:05,019 p=6706 u=tjuberg |  ncclient manager object created successfully 
2017-05-12 08:46:05,020 p=6706 u=tjuberg |  connection established to xxx.xxx.xxx.xxx in 0:00:00.590981 
2017-05-12 08:46:05,427 p=6706 u=tjuberg |  incoming request accepted on persistent socket 
2017-05-12 08:46:05,427 p=6706 u=tjuberg |  socket operation is CONTEXT 
2017-05-12 08:46:05,428 p=6706 u=tjuberg |  socket operation is EXEC 
2017-05-12 08:46:05,429 p=6706 u=tjuberg |  socket operation completed with rc 255 
2017-05-12 08:46:05,493 p=6697 u=tjuberg |  open_session() returned 255 b'' b'Traceback (most recent call last):\n  
File "/home/tjuberg/.local/badger-ansible/bin/ansible-connection", line 199, in run\n
(rc, stdout, stderr) = self.conn.exec_command(cmd)\n
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/__init__.py", line 52, in wrapped\n
return func(self, *args, **kwargs)\n
File "/home/tjuberg/.local/badger-ansible/lib/ansible/plugins/connection/netconf.py", line 113, in exec_command\n
req = to_ele(request)\n
File "/home/tjuberg/.local/lib/python3.6/site-packages/ncclient/xml_.py", line 109, in to_ele\n
return x if etree.iselement(x) else etree.fromstring(x.encode(\'UTF-8\'), parser=parser)\n
AttributeError: \'bytes\' object has no attribute \'encode\'\n' 
2017-05-12 08:46:05,495 p=6697 u=tjuberg |  fatal: [trd-gw]: FAILED! => { 
    "changed": false, 
    "failed": true, 
    "msg": "unable to open shell. Please see: https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell", 
    "rc": 255 
} 
2017-05-12 08:46:05,496 p=6697 u=tjuberg |  	to retry, use: --limit @/home/tjuberg/work/gitlab/bachelor_ansible/junos_facts.retry 

2017-05-12 08:46:05,497 p=6697 u=tjuberg |  PLAY RECAP ********************************************************************************************************************* 
2017-05-12 08:46:05,497 p=6697 u=tjuberg |  trd-gw                     : ok=0    changed=0    unreachable=0    failed=1    
2017-05-12 08:46:35,429 p=6706 u=tjuberg |  shutting down control socket, connection was active for 0:00:31.000909 secs 
2017-05-12 08:46:35,430 ncclient.operations.rpc Requesting 'CloseSession' 
```

ncclient  0.5.3 for reference.

**Comment by abadger** [comment]
merged to devel for inclusion in 2.4.x

@TJuberg I'm going to create another PR to start addressing the junos incompatibilities.  I'll CC you on that as soon as I get it posted.

**Pull request merged**